Dockerfile

```diff
@@ -1,16 +1,27 @@
 # ============ Build Stage ============
-FROM rust:1.74-bookworm as build
-
+FROM lukemathwalker/cargo-chef:latest-rust-1 AS chef
 WORKDIR /typst-bot
 
+FROM chef AS planner
+
 # Compilation requires only the source code.
 COPY Cargo.toml Cargo.lock ./
```

Comment:
Since you copying all code before `cargo chef`, I suspect that docker will rebuild everything

Reply:
Hmm, maybe I'm wrong, `cargo chef` docs suggest the same

Reply:
you need the source code to generate the recipe.json

Reply:
I was wondering the same thing. (BTW, I thought that it said `COPY . .` just a few hours ago.) But I did a research, and it turns out everything is very simple: the planner stage will get invalidated when your source code changes, yes. But! If the dependencies etc. are the same, then it will result in the **exact same** `recipe.json` which in turn _won't_ invalidate the builder stage cache, because its dependencies (`recipe.json`) didn't change at all. Therefore, the most costly stage will be cached.

I didn't understand this (fully) just by reading https://github.com/LukeMathWalker/cargo-chef/issues/256#issuecomment-1954597650, but then it hit me after reading (a ton of Docker docs, that I already almost fully knew, and) this: https://www.lpalmieri.com/posts/fast-rust-docker-builds/.